File: archive/v2-csharp-avalonia/src/Spanova.Rules/RuleEngine.cs
using Spanova.Core.Model;

namespace Spanova.Rules;

/// <summary>Traceable outcome of running every configured rule against one alternative.</summary>
public sealed record RuleEngineResult(bool IsFeasible, IReadOnlyList<(IEngineeringRule Rule, RuleEvaluation Evaluation)> Evaluations)
{
    public IEnumerable<string> FailureReasons =>
        Evaluations.Where(e => !e.Evaluation.Passed).Select(e => $"[{e.Rule.RuleId}] {e.Evaluation.Reason}");
}

/// <summary>
/// Runs a configured set of <see cref="IEngineeringRule"/>s against a
/// bridge alternative (spec section 20, P04: "Generated alternatives can
/// be accepted/rejected with traceable rule results").
///
/// The rule set defaults to empty because no engineering rule has been
/// approved yet (spec section 22) - every alternative is therefore
/// "feasible" vacuously until the engineer supplies the first real rule.
/// This is P04's actual, intended state: the mechanism works and is
/// tested (see RuleEngineTests, which uses deliberately fake rules to
/// prove the plumbing), but it has no engineering content.
/// </summary>
public sealed class RuleEngine
{
    private readonly IReadOnlyList<IEngineeringRule> _rules;

    public RuleEngine(IReadOnlyList<IEngineeringRule>? rules = null)
    {
        _rules = rules ?? [];
    }

    public RuleEngineResult Evaluate(BridgeAlternative alternative)
    {
        var evaluations = _rules
            .Select(rule => (Rule: rule, Evaluation: rule.Evaluate(alternative)))
            .ToList();

        var isFeasible = evaluations.All(e => e.Evaluation.Passed);

        return new RuleEngineResult(isFeasible, evaluations);
    }
}

File: backend/rules-engine/src/main/java/com/kopruq/rules/RuleEngine.java
package com.kopruq.rules;

import com.kopruq.bridgecore.Bridge;
import com.kopruq.bridgecore.BridgeAlternative;
import com.kopruq.bridgecore.DesignSpace;

import java.util.List;

/**
 * Evaluates a set of {@link EngineeringRule}s against a bridge
 * alternative (spec section 11, P04). Ships with an empty rule set by
 * default - see {@link com.kopruq.rules package-info}: no rule content
 * exists yet, and none should be added without the engineer's explicit
 * approval.
 */
public final class RuleEngine {

    private final List<EngineeringRule> rules;

    public RuleEngine() {
        this(List.of());
    }

    public RuleEngine(List<EngineeringRule> rules) {
        this.rules = List.copyOf(rules);
    }

    public RuleEngineResult evaluate(BridgeAlternative alternative, Bridge bridge, DesignSpace designSpace) {
        var outcomes = rules.stream()
                .map(rule -> new RuleOutcome(
                        rule.id(),
                        rule.description(),
                        rule.evaluate(alternative, bridge, designSpace)))
                .toList();

        boolean feasible = outcomes.stream().allMatch(outcome -> outcome.evaluation().passed());

        return new RuleEngineResult(feasible, outcomes);
    }
}
